### PR TITLE
Detect zlib and enable DNG deflate codec support

### PR DESCRIPTION
**Pull request by kmilos**

### CODE REVIEW

**Comment by letmaik** [comment]
Thanks for this. I saw there are two functions in libraw: lossless_dng_load_raw and deflate_dng_load_raw. You're enabling the latter with your patch, but the PR title says "lossless codec". Can you clarify this? Are both lossless and one of them is using deflate compression as well?

**Comment by kmilos** [comment]
This enables only the implementation of deflate_dng_load_raw() via zlib - indeed using the DEFLATE lossless compression method, so I named it as such as opposed to the already enabled lossy one via the JPEG library. To make things worse, I now see there is also lossless_dng_load_raw() and lossless_jpeg_load_raw(), presumably both referring to the lossless JPEG compression implemented differently?

More details on various options are in the DNG 1.4 spec, "Compression" section on p19.

I'm not opposed to changing the term to "deflate" if you think it's confusing as it stands.

**Comment by letmaik** [comment]
My confusion would be that libraw always supports lossless DNG (via lossless JPEG (if JPEG is enabled) or uncompressed), so this is just another variant of lossless. I would make it more explicit, and since deflate implies lossless we could name it simply `LIBRAW_USE_DNGDEFLATECODEC` which also matches function naming in libraw.

File: CMakeLists.txt [hunk 1]
```diff
@@ -486,6 +505,10 @@ IF(LCMS_SUPPORT_CAN_BE_COMPILED)
     TARGET_LINK_LIBRARIES(raw ${LCMS2_LIBRARIES})
 ENDIF()
 
+IF(ZLIB_FOUND)
+    TARGET_LINK_LIBRARIES(raw ${ZLIB_LIBRARY})
```
Comment (letmaik) [inline CMakeLists.txt:509]: Please change to ZLIB_LIBRARIES. See https://cmake.org/cmake/help/v3.14/module/FindZLIB.html.

File: CMakeLists.txt [hunk 2]
```diff
@@ -153,6 +153,19 @@ ENDIF()
...
 MACRO_BOOL_TO_01(LCMS_SUPPORT_CAN_BE_COMPILED LIBRAW_USE_LCMS)
 
+# zlib library check
+
+FIND_PACKAGE(zlib)
+
+# Flag to use zlib with LibRaw DNG deflate codec
+IF(ZLIB_FOUND)
+    ADD_DEFINITIONS(-DUSE_ZLIB)
+    INCLUDE_DIRECTORIES(${ZLIB_INCLUDE_DIR})
```
Comment (letmaik) [inline CMakeLists.txt:163]: Please change to ZLIB_INCLUDE_DIRS, see https://cmake.org/cmake/help/v3.14/module/FindZLIB.html.

File: CMakeLists.txt [hunk 3]
```diff
@@ -530,6 +553,10 @@ IF(LCMS_SUPPORT_CAN_BE_COMPILED)
     TARGET_LINK_LIBRARIES(raw_r ${LCMS2_LIBRARIES})
 ENDIF()
 
+IF(ZLIB_FOUND)
+    TARGET_LINK_LIBRARIES(raw_r ${ZLIB_LIBRARY})
```
Comment (letmaik) [inline CMakeLists.txt:557]: Same here.

**Comment by letmaik** [comment]
@alextutubalin This PR was not finished yet, see my comments. Please only merge PRs in this repo if you really have a good understanding of them. As far as I remember you don't want to support CMake officially and have handed over CMake maintenance to myself.

**Comment by letmaik** [comment]
@kmilos Sorry about this. Can you create another PR with the remaining fixes? I can also do it myself if you prefer.

**Comment by alextutubalin** [comment]
@letmaik, sorry for that: in my opinion, changing description to DNG/deflate was enough.

Is there any way to completely transfer control to you?

**Comment by LibRaw** [comment]
I've initiated transfer to you :)

**Comment by letmaik** [comment]
@alextutubalin I think it would be better to keep it under the LibRaw organisation so that others can find it easily. I already have contributor access to this repository and can merge PRs. I think that's good enough, no? You don't have to deal with anything really. Have a look at the [README](https://github.com/LibRaw/LibRaw-cmake/blob/master/README.md) again and see if that makes everything clear regarding that this is not officially supported by the libraw developers/authors.

**Comment by kmilos** [comment]
@letmaik Please feel free to do so. I'm not a Cmake expert, and my FindZlib.cmake had these definitions...

**Comment by letmaik** [comment]
@kmilos OK no worries, I'll make the change. The variables you used also work but they are kind of internal/private variables. That's why they're not mentioned in the CMake docs. Thanks again for your contribution! I'll pick this up in the next release of my libraw Python bindings as well (https://github.com/letmaik/rawpy) :)

**Comment by kmilos** [comment]
@letmaik Thanks, rawpy was actually the reason I pushed the change 😉

**Comment by LibRaw** [comment]
@letmaik, the problem is E-mail I receive when someone contributed here.

It is hard to hold on and do not press  merge button :)

**Comment by letmaik** [comment]
@alextutubalin Ah, I understand now, haha. Actually, GitHub has a feature where you can unsubscribe from a repository completely. Just click on "Unwatch" at the top right and select "Not watching". Maybe that's a good compromise?

**Pull request merged**